src/ActionList/List.tsx

```diff
@@ -0,0 +1,69 @@
+import {Group, GroupProps} from './Group'
+import {Item, ItemProps} from './Item'
+import React from 'react'
+import {Divider} from './Divider'
+import {Header, HeaderProps} from './Header'
+import styled from 'styled-components'
+import {get} from '../constants'
+
+type Flatten<T extends unknown> = T extends (infer U)[] ? U : never
+
+interface UngroupedListProps {
+  items: ItemProps[]
+  renderItem?: (props: ItemProps) => JSX.Element
+}
+function isUngroupedListProps(props: ListProps): props is UngroupedListProps {
+  return typeof props === 'object' && props !== null && !('groupMetadata' in props)
+}
+
+interface GroupedListProps extends UngroupedListProps {
+  groupMetadata: (GroupProps & {groupId: number; header?: HeaderProps})[]
```

Comment:
Can `groupId` be changed to a string? I think it would make for more readable consumer code to have group names instead of numeric IDs. If so, consider changing `groupId` to `groupName`.

Reply:
Fixed in https://github.com/primer/components/pull/1078/commits/f95e1a0aa9189ced510cd408bf1e6bff83c6a90c!